packages/grid/x-grid/src/tests/rows.XGrid.test.tsx

```diff
@@ -249,4 +251,97 @@ describe('<XGrid /> - Rows ', () => {
       expect(getColumnValues()).to.deep.equal(['Apple', 'Atari']);
     });
   });
+
+  describe('virtualization', () => {
+    let apiRef: GridApiRef;
+    const TestCaseVirtualization = (
+      props: Partial<XGridProps> & { nbRows?: number; nbCols?: number; height?: number },
+    ) => {
+      apiRef = useGridApiRef();
+      const data = useData(props.nbRows || 100, props.nbCols || 10);
+
+      return (
+        <div style={{ width: 300, height: props.height || 300 }}>
+          <XGrid apiRef={apiRef} columns={data.columns} rows={data.rows} {...props} />
+        </div>
+      );
+    };
+
+    it('Rows should not be virtualized when the number of rows fit in the viewport', () => {
+      const headerHeight = 50;
+      const rowHeight = 50;
+      const maxRowsNotVirtualised = (300 - headerHeight) / rowHeight;
+      render(
+        <TestCaseVirtualization
+          nbRows={maxRowsNotVirtualised}
+          hideFooter
+          headerHeight={headerHeight}
+          rowHeight={rowHeight}
+        />,
+      );
+
+      const isVirtualized = apiRef!.current!.getState().containerSizes!.isVirtualized;
+      expect(isVirtualized).to.equal(false);
+    });
+
+    it('Rows should be virtualized when at least 2 rows are outside the viewport', () => {
+      const headerHeight = 50;
+      const rowHeight = 50;
+      const maxRowsNotVirtualised = (300 - headerHeight) / rowHeight;
+      render(
+        <TestCaseVirtualization
+          nbRows={maxRowsNotVirtualised + 1}
+          hideFooter
+          headerHeight={headerHeight}
+          rowHeight={rowHeight}
+        />,
+      );
+
+      let isVirtualized = apiRef!.current!.getState().containerSizes!.isVirtualized;
+      expect(isVirtualized).to.equal(false);
+
+      render(
+        <TestCaseVirtualization
+          nbRows={maxRowsNotVirtualised + 2}
+          hideFooter
+          headerHeight={headerHeight}
+          rowHeight={rowHeight}
+        />,
+      );
+
+      isVirtualized = apiRef!.current!.getState().containerSizes!.isVirtualized;
+      expect(isVirtualized).to.equal(true);
+    });
+
+    it('should render last row when scrolling to the bottom', () => {
+      const { container } = render(<TestCaseVirtualization nbRows={996} hideFooter height={600} />);
+      const totalHeight = apiRef!.current!.getState().containerSizes?.totalSizes.height;
+
+      const gridWindow = container.querySelector('.MuiDataGrid-window');
+      const renderingZone = container.querySelector('.MuiDataGrid-renderingZone');
+      gridWindow.scrollTop = totalHeight;
+      gridWindow.dispatchEvent(new Event('scroll'));
+
+      const lastCell = container.querySelector(
+        '.MuiDataGrid-row:last-child .MuiDataGrid-cell:first-child',
```

Comment:
Could we always use the a11y features first (the role instead of the class names)?
```suggestion
        '[role="row"]:last-child [role="cell"]:first-child',
```